openhtf/exe/__init__.py

```diff
@@ -122,40 +137,99 @@ def _RunTestLoop(self):
       # Top level steps required to run a single iteration of the Test.
       _LOG.info('Starting test %s', self.test.code_info.name)
 
-      # Save exit_stack on self so we can access it when Stop() is called.
-      self._current_exit_stack = exit_stack
-      exit_stack.callback(lambda: setattr(self, '_current_exit_stack', None))
-
+      # Any access to self._exit_stack must be done while holding this lock.
+      with self._lock:
+        # Initial setup of exit stack and final cleanup of attributes.
+        self._exit_stack = exit_stack
+        exit_stack.callback(self._ResetAttributes)
+        exit_stack.callback(self._MaybeJoinOutputThread)
+
+      # Wait here until the test start trigger returns a DUT ID.  Don't hold
+      # self._lock while we do this, or else calls to Stop() will deadlock.
+      # Create plugs while we're here because that may also take a while and
+      # we don't want to hold self._lock while we wait.
+      dut_id = self._WaitForTestStart(suppressor)
+      self._status = self.FrameworkStatus.INITIALIZING
       plug_manager = self._MakePlugManager(suppressor)
-      executor = self._MakePhaseExecutor()
 
+      with self._lock:
+        if not self._exit_stack:
+          # We shouldn't get here, but just in case something went weird with a
+          # call to Stop() and we ended up resuming execution here but the
+          # exit stack was already cleared, bail.  Try to tear down plugs on a
+          # best-effort basis.
+          plug_manager.TearDownPlugs()
+          raise TestStopError('Test Stopped.')
+        exit_stack.callback(plug_manager.TearDownPlugs)
+
+        # Perform initialization of some top-level stuff we need.
+        self._test_state = self._MakeTestState(dut_id, plug_manager, suppressor)
+        executor = self._MakePhaseExecutor(exit_stack, suppressor)
+
+      # Everything is set, set status and begin test execution.  Note we don't
+      # protect this with a try: block because the PhaseExecutor handles any
+      # exceptions from test code.  Any exceptions here are caused by the
+      # framework, and we probably want them to interrupt framework state
+      # changes (like the transition to FINISHING).
       self._status = self.FrameworkStatus.EXECUTING
       suppressor.failure_reason = 'Failed to execute test.'
       self._ExecuteTestPhases(executor)
       self._status = self.FrameworkStatus.FINISHING
-      self.test.OutputTestRecord(self._test_state.GetFinishedRecord())
 
-  def _MakePlugManager(self, suppressor):
-    """Perform some initialization and create a PlugManager."""
+      # Output the test record now that we're done executing.
+      self._output_thread = self._MakeAndStartOutputThread()
```

Comment:
Sounds good! What facelift for the output mechanism? I sort of thought of
making the TestExecutor return the TestRecord at the end and have the Test
class do the outputting (after the while loop got pulled out too) since it
has the callbacks anyway.
On Mar 27, 2016 8:00 PM, "grybmadsci" <notifications@github.com> wrote:

> In openhtf/exe/__init__.py
> <https://github.com/google/openhtf/pull/200#discussion_r57541267>:
>
> >
> > -  def _MakePlugManager(self, suppressor):
> > -    """Perform some initialization and create a PlugManager."""
> > +      # Output the test record now that we're done executing.
> > +      self._output_thread = self._MakeAndStartOutputThread()
>
> Hm, yeah that's a good point, I was considering the same thing with
> providing access to the record logger in BasePlug, it's kind of awkward if
> the output callbacks get called before plug tear down (and it's just racey
> if they happen in parallel).
>
> I had done it this way because in my case i'm using the output callback to
> display result via GUI and block until user closes the dialog, but that's
> actually me conflating test stop trigger and output, so i'll fix that on my
> end, and switch this to be less stupid and just tear down plugs *then*
> output the record.
>
> when i do that i think i'll remove the separate thread business for now.
> it'll likely come back once the output mechanism gets a facelift, but i
> think it's overkill for the current state, what do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/google/openhtf/pull/200/files/f839302d9ce4859f358d462c276f1980b94d4eaa#r57541267>
>